src/ember-app.js

```diff
@@ -227,6 +237,28 @@ class EmberApp {
     return app;
   }
 
+  /**
+   * @private
+   *
+   * Get the new sandbox off if it is being created, otherwise create a new one on demand.
+   * The later is needed when the current request hasn't finished or wasn't build with sandbox
+   * per request turned on and a new request comes in.
+   *
+   */
+  async _getNewApplicationInstance() {
+    let app;
+
+    if (this._pendingNewApplicationInstance) {
+      app = await this._pendingNewApplicationInstance;
+      this._pendingNewApplicationInstance = undefined;
```

Comment:
I think we need to reset this to undefined _before_ waiting (otherwise we have a race condition and sending many pending requests at the same time might force the same sandbox to be used accidentally).